openapi.yml

```diff
@@ -0,0 +1,105 @@
+openapi: 3.0.3
+info:
+  version: 0.0.1
+  title: Order Book API
+servers:
+- url: http://localhost:8080
+  description: Local
+paths:
+  /api/v1/order:
+    post:
+      summary: Create a new order.
+      responses:
+        201:
+          description: Order has been accepted.
+      requestBody:
+        description: The order to create.
+        required: true
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/Order"
+  /api/v1/orders:
+    get:
+      summary: Get existing orders.
+      # TODO: Later we could add filters through the url query to only get orders matching specific tokens etc.
+      responses:
+        200:
+          description: existing orders
+          content:
+            application/json:
+              schema:
+                type: array
+                items:
+                  $ref: "#/components/schemas/Order"
+  /api/v1/nonce:
+    get:
+      summary: The nonce to pick for a new order.
+      description: The best nonce to pick depends on how much time there is left in the current settlement according to the operator. This time is token-pair specific which is why the sell and buy tokens must be supplied.
+      parameters:
+        - name: sellToken
+          in: query
+          schema:
+            $ref: "#/components/schemas/Address"
+        - name: buyToken
+          in: query
+          schema:
+            $ref: "#/components/schemas/Address"
+      responses:
+        200:
+          description: the nonce
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Nonce"
+components:
+  schemas:
+    Address:
+      description: Ethereum 40 byte address encoded as a hex.
+      type: string
+      example: "0x6810e776880c02933d47db1b9fc05908e5386b96"
+    TokenAmount:
+      description: Amount of a token. uint112 encoded in decimal.
+      type: string
+      example: "1234567890"
+    Nonce:
+      description: Nonce of the batch in which the order is valid. uint32.
+      type: integer
+    Order:
```

Comment:
An open question to me is whether the order should also contain a `createdAt` field. The spec says we should settle orders the latest when they are 1 minute old so we need to have this information somewhere.

1. We could let the user (really the frontend) supply this value.
2. We could make the orderbook choose the value and return it when getting existing orders.
3. We could ignore this in the orderbook and make the solver keep track of it.

1 has the disadvantage that the value can be tempered with like setting it in the past so that the order appears older than it really is.
3 makes the orderbook slightly simpler but seems very finicky in the solver code because we would need to keep track of the first time we see a specific order from the orderbook.
2 seems like the best choice.

Reply:
Yes, I fully agree. 2 is the way to go